Content.Server/Roles/RoleTimerManager.cs

```diff
@@ -0,0 +1,177 @@
+        {
+            return _cachedPlayerData.TryGetValue(id, out var timers) && timers.Initialized;
+        }
+
+        /// <summary>
+        /// Puts all relevant database information for a player into the cache.
+        /// </summary>
+        private async Task CachePlayerRoles(NetUserId player, PlayerRoleTimers timers)
+        {
+            if (timers.Initialized) return;
+            var roleTimers = await _db.GetRoleTimers(player);
+            var playtime = await _db.GetOverallPlayTime(player);
+
+            foreach (var timer in roleTimers)
+            {
+                await timers.AddPlaytimeForRole(timer.Role, timer.TimeSpent);
+            }
+
+            await timers.AddOverallPlaytime(playtime);
+            timers.Initialized = true;
+        }
+
+        public async void AddTimeToRole(NetUserId id, string role, TimeSpan time, bool dbSave = true)
+        {
+            if (!_cachedPlayerData.TryGetValue(id, out var timers))
+                return;
+
+            var totalTime = await timers.AddPlaytimeForRole(role, time);
+
+            if (!dbSave) return;
+
+            var query = await _db.CreateOrGetRoleTimer(id, role);
+            await _db.SetRoleTime(query.Id, totalTime);
+        }
+
+        public async void AddTimeToOverallPlaytime(NetUserId id, TimeSpan time, bool dbSave = true)
+        {
+            if (!_cachedPlayerData.TryGetValue(id, out var timers))
+                return;
+
+            var playtime = await timers.AddOverallPlaytime(time);
+            if (!dbSave) return;
+            await _db.SetOverallPlayTime(id, playtime);
+        }
+
+        public async Task<TimeSpan> GetOverallPlaytime(NetUserId id)
+        {
+            if (!_cachedPlayerData.TryGetValue(id, out var timer)) return TimeSpan.Zero;
+            return await timer.GetOverallTime();
+        }
+
+        public async Task<Dictionary<string, TimeSpan>> GetRolePlaytimes(NetUserId id)
+        {
+            if (!_cachedPlayerData.TryGetValue(id, out var timer)) return new Dictionary<string, TimeSpan>();
+            return await timer.GetRoleTimers();
+        }
+
+        public async Task<TimeSpan> GetPlayTimeForRole(NetUserId id, string role)
+        {
+            if (!_cachedPlayerData.TryGetValue(id, out var timers))
+            {
+                return TimeSpan.Zero;
+            }
+
+            return await timers.GetPlaytimeForRole(role);
+        }
+
+        /// <summary>
+        /// Role timers for a particular player.
+        /// </summary>
+        private sealed class PlayerRoleTimers
+        {
+            /// <summary>
+            /// Have we already retrieved our data from the DB?
+            /// </summary>
+            public bool Initialized = false;
+
+            private readonly Dictionary<string, TimeSpan> _roleTimers = new();
+            private TimeSpan _overallPlaytime = TimeSpan.Zero;
+
+            private readonly SemaphoreSlim _semaphore = new(1);
```

Comment:
So two things here:
1. All the actual logic in the critical sections is so small you're wasting more CPU cycles using asynchronous semaphore awaits than if you'd just use `lock()`. Something which would also avoid all the asynchronous leakage into the consuming code.
2. This synchronization logic isn't even necessary because the sync context ensures awaits resume on the main thread.
3. For future advice: if you really really need to lock something, you should still have put the synchronization externally from these internal methods. When loading role timers you are constantly grabbing and releasing the lock. Not good.